expat/lib/xmlparse.c

```diff
@@ -5438,6 +5489,31 @@ processInternalEntity(XML_Parser parser, ENTITY *entity,
       parser->m_freeInternalEntities = openEntity;
     }
   }
+
+  if (!(parser->m_options & XML_OPTION_HUGE_ENTITES)) {
+    XML_Index index;
+    XML_Index limit;
+    parser->m_entityRecursionLevel--;
+    if (entity->textLen > XML_ENTITY_EXPANSION_SIZE) {
+      /* Current entity is too large */
+      return XML_ERROR_ENTITY_TOO_LARGE;
+    }
+    /* Entity expension length isn't equal to output length. For nested
+     * entities, it records length of each expansion level and not sum of
+     * expanded text. */
+    parser->m_entityExpansionLen += entity->textLen;
+    index = XML_GetCurrentByteIndex(parser);
+    limit = index * XML_ENTITY_EXPANSION_RATIO;
+    if (limit < index) {
```

Comment:
To my best knowlegde signed integer overflow is undefined behavior, hence we'd need to prevent overflow before it even happened. Both `index` and `XML_ENTITY_EXPANSION_RATIO` are signed as far as I can see.

Reply:
We could try division instead to bypass that issue. On a chalkboard I'd write in order:

1. Right now we have:
  `parser->m_entityExpansionLen > limit`.
1. Inlined, that's:
  `parser->m_entityExpansionLen > index * XML_ENTITY_EXPANSION_RATIO`.
1. If we devide both sides by `XML_ENTITY_EXPANSION_RATIO` we get:
  `parser->m_entityExpansionLen / XML_ENTITY_EXPANSION_RATIO > index`.
1. If we take the floor effect of integer division into account and correct for it to accept the same range we get:
   `(parser->m_entityExpansionLen + (XML_ENTITY_EXPANSION_RATIO - 1)) / XML_ENTITY_EXPANSION_RATIO > index`.

Does that look about right?

That could be one way. Another way could be leveraging `(XML_Size)(-1)` since that type is unsigned. I'd need to think about that more.